feat: ooo calibration

CarinaWolli opened the pull request:
## What does this PR do?

Add OOO calibration for weighted round robin. It take into account OOO entries and full day busy events from google calendar. 

- Fixes CAL-4497

## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Checklist

<!-- Remove bullet points below that don't apply to you -->

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

Commit:
first version of ooo calibration

emrysal commented on packages/lib/server/getLuckyUser.ts:
```diff
@@ -570,6 +671,63 @@ async function fetchAllDataNeededForCalculations<
+    },
+  });
+
+  const groupedOOOEntries = new Map<number, { start: Date; end: Date }[]>();
```
maybe `oooEntriesGroupedByUserId` to be more explicit about what it's grouping - saves scanning the further code.

emrysal commented on packages/lib/server/getLuckyUser.ts:
```diff
@@ -455,6 +526,34 @@ export function getLuckyUser_requiresDataToBePreFetched<
+  return mergedRanges;
+}
+
+function isFullDayEvent(date1: Date, date2: Date) {
```
```suggestion
function isFullDayEvent(date1: Date, date2: Date) {
  const MILLISECONDS_IN_A_DAY = 24 * 60 * 60 * 1000;
  const difference = Math.abs(date1.getTime() - date2.getTime());

  // Check if the time difference is approximately one full day
  return Math.abs(difference - MILLISECONDS_IN_A_DAY) < 1000;
```

Just to prevent a case where we're like 365ms off and it doesn't see it as a full day - decrease precision.

emrysal commented on packages/lib/server/getLuckyUser.ts:
```diff
@@ -455,6 +526,34 @@ export function getLuckyUser_requiresDataToBePreFetched<
   };
 }
 
+function mergeOverlappingRanges(ranges: { start: Date; end: Date }[]): { start: Date; end: Date }[] {
```
This function is best extracted

emrysal replied:
@calcom/lib/date-ranges

CarinaWolli committed:
decrease precision when checking for full day event

Co-authored-by: Alex van Andel <me@alexvanandel.com>

Commit:
improve variable naming

Commit:
don't include hosts created after OOO

Commit:
fix failing tests

Commit:
fix all tests

Commit:
add test for ooo calibration

Commit:
first implementation to fix full day busy event reaching the current month

Commit:
allow mutiple day events

Commit:
make getAvailabilityWithTimeZones optional

Commit:
fix type error

Commit:
fix type errors

Commit:
fix tests

Commit:
fix type error

Commit:
fix test

Commit:
code clean up

Commit:
remove time adjustment from map

Commit:
code clean up

Commit:
fix tests

hbjORbj commented on packages/core/CalendarManager.ts:
```diff
@@ -199,20 +200,24 @@ const getMonths = (dateFrom: string, dateTo: string): string[] => {
 };
 
 export const getBusyCalendarTimes = async (
```
did you make sure to update every instance of this function?

emrysal replied:
yeah this would've been caught out by type checks

hbjORbj committed:
Merge remote-tracking branch 'origin/main' into feat/dynamic-ooo-calibration

hbjORbj commented on packages/app-store/googlecalendar/lib/CalendarService.ts:
```diff
@@ -515,18 +513,90 @@ export default class GoogleCalendarService implements Calendar {
+    }
+
+    async function getCalIdsWithTimeZone() {
+      const cals = await getAllCalendars(calendar, ["id", "timeZone"]);
```
`lib/google` is deleted, so you need to use `this.getAllCalendars` instead

hbjORbj replied:
I merged main into your branch and updated this

hbjORbj commented on packages/app-store/googlecalendar/lib/CalendarService.ts:
```diff
@@ -515,18 +513,90 @@ export default class GoogleCalendarService implements Calendar {
+
+    try {
+      const calIdsWithTimeZone = await getCalIdsWithTimeZone();
+      const calIds = calIdsWithTimeZone.map((calIdWithTimeZone) => calIdWithTimeZone.id);
```
```suggestion
      const calIds = calIdsWithTimeZone.map((calIdWithTimeZone) => ({id: calIdWithTimeZone.id}));
```

hbjORbj commented on packages/app-store/googlecalendar/lib/CalendarService.ts:
```diff
@@ -515,18 +513,90 @@ export default class GoogleCalendarService implements Calendar {
+      const freeBusyData = await this.getCacheOrFetchAvailability({
+        timeMin: dateFrom,
+        timeMax: dateTo,
+        items: calIds.map((id) => ({ id })),
```
```suggestion
        items: calIds,
```

hbjORbj commented on packages/app-store/googlecalendar/lib/CalendarService.ts:
```diff
@@ -515,18 +513,90 @@ export default class GoogleCalendarService implements Calendar {
+      });
+      if (!freeBusyData) throw new Error("No response from google calendar");
+
+      const freeBusyDataWithTimeZone = freeBusyData.map((freeBusy) => {
+        let timeZone = "";
+        const calIdWithTimeZone = calIdsWithTimeZone.find(
+          (calIdWithTimeZone) => calIdWithTimeZone.id === freeBusy.id
+        );
+        timeZone = calIdWithTimeZone?.timeZone || "";
+        return {
+          start: freeBusy.start,
+          end: freeBusy.end,
+          timeZone,
+        };
+      });
```
We can avoid the nested loop in this way:
```jsx
const timeZoneMap = new Map(
  calIdsWithTimeZone.map(cal => [cal.id, cal.timeZone])
);

const freeBusyDataWithTimeZone = freeBusyData.map(freeBusy => ({
  start: freeBusy.start,
  end: freeBusy.end,
  timeZone: timeZoneMap.get(freeBusy.id) || ""
}));
```

emrysal commented on packages/app-store/googlecalendar/lib/CalendarService.ts:
```diff
@@ -1,5 +1,4 @@
 /* eslint-disable @typescript-eslint/no-explicit-any */
-import type { Prisma } from "@prisma/client";
 import type { calendar_v3 } from "googleapis";
```
This has been replaced also, check with @hbjORbj - should be more performant henseforth.

hbjORbj committed:
getAllCalendars -> this.getAllCalendars

emrysal reviewed (changes requested):
Some nits, also looks like getOrderedListOfLuckyUsers is out of order 🤔  - soft request changes, feel free to disregard when you deem it ready.

Commit:
fix typo

Commit:
extract mergeOverlappingRanges

Commit:
code improvements from feedback

Commit:
fix getCalIdsWithTimeZone

CarinaWolli commented:
@emrysal @hbjORbj Thank you for your review, I addressed all your comments + fixed one more bug I found https://github.com/calcom/cal.com/pull/17864#discussion_r1882349215

> also looks like getOrderedListOfLuckyUsers is out of order

I have seen this before, let's fix this seperatly

emrysal reviewed (dismissed):
Looks good to me, very happy to merge.

Commit:
Merge branch 'main' into feat/dynamic-ooo-calibration

Commit:
add changed lost in merge

emrysal reviewed (approved):
Still happy with this 😄
